src/net_doh.c

```diff
@@ -0,0 +1,1140 @@
+                                        sizeof(hdrs) / sizeof(hdrs[0]),
+                                        NULL,
+                                        self->http2->stream);
+    if (stream_id < 0) {
+        perf_log_fatal("Failed to submit HTTP2 request: %s", nghttp2_strerror(stream_id));
+    }
+
+    self->http2->stream->stream_id = stream_id;
+
+    debugx("qid: %d, stream_id: %d", self->qid, stream_id);
+
+    ret = nghttp2_session_send(self->http2->session);
+    if (ret < 0) {
+        perf_log_warning("nghttp2_session_send failed: %d", ret);
+        return -1;
+    }
+
+    return 0;
+}
+
+static int _submit_dns_query_post(struct perf_net_socket* sock, const void* buf, size_t len)
+{
+    return 0;
+}
+
+static http2_stream_t* http2_stream_init(struct URI* uri)
+{
+    http2_stream_t *stream_data = calloc(1, sizeof(http2_stream_t));
+
+    stream_data->path = calloc(1, uri->pathlen + 1);
+    memcpy(stream_data->path, uri->path, uri->pathlen);
+    stream_data->path[uri->pathlen] = '\0';
+
+    stream_data->authority = calloc(1, uri->hostportlen + 1);
+    memcpy(stream_data->authority, uri->hostport, uri->hostportlen);
+    stream_data->authority[uri->hostportlen] = '\0';
+
+    return stream_data;
+}
+
+static void http2_stream_free(http2_stream_t *stream_data) {
+    free(stream_data->path);
+    free(stream_data->authority);
+    free(stream_data);
+}
+
+http2_session_t* http2_session_init()
+{
+    http2_session_t *session_data = calloc(1, sizeof(http2_session_t));
+    memset(session_data, 0, sizeof(http2_session_t));
+    return session_data;
+}
+
+static void http2_session_free(struct perf_net_socket* sock) 
+{
+    if (self->ssl) {
+        SSL_shutdown(self->ssl);
+    }
+    
+    nghttp2_session_del(self->http2->session);
+    self->http2->session = NULL;
+
+    if (self->http2->stream) {
+        http2_stream_free(self->http2->stream);
+        self->http2->stream = NULL;
+    }
+
+    free(self->http2);
+    self->http2 = NULL;
+}
+
+/* nghttp2 callbacks */
+
+// TODO: remove - we handle TLS read upstream
+static ssize_t _http2_recv_cb(nghttp2_session* session,
+                          const uint8_t* data, 
+                          size_t len, 
+                          int flags,
+                          void* user_data)
+{
+    // TODO:
+    debugx("http2_recv_cb - len: %d", len);
+    return 0;
+}
+
+static ssize_t _http2_send_cb(nghttp2_session* session, 
+                                    const uint8_t* data, 
+                                    size_t length, 
+                                    int flags, 
+                                    void* user_data)
+{
+    ssize_t n;
+    (void)session;
+    (void)flags;
+
+    struct perf_net_socket *sock = (struct perf_net_socket *)user_data;
+    (void)session;
+    (void)flags;
+
+    if (!self->is_ready) {
+        return NGHTTP2_ERR_CALLBACK_FAILURE;
+    }
+
+    n = SSL_write(self->ssl, data, length);
```

Comment:
no lock on SSL context, is this callback protected?

Reply:
on every write operation to the http2 layer i.e. nghttp2_submit_*, we follow with nghttp2_session_send() which triggers the callback. We also do _session_send() in sockready() to drive the exec of http2 i/o. In both cases we are within the respective mutex lock/unlock cycle, so I believe it is protected.